text/0000-use.md

```diff
@@ -0,0 +1,379 @@
+implementation seems to be obviously breaking the intended constraints on the
+`Use` trait.  Such a lint would be best-effort only, and could always be marked
+as `allow` by a crate, but could help to discourage such implementations.
+
+We may want to add a clippy or rustc lint for calls to `.clone()` that could
+use `.use` instead. This would help the remaining calls to `.clone()` stand out
+as "expensive" clones. (Such a lint would need to take MSRV into account before
+making such a suggestion; clippy already has such a mechanism and rustc may
+gain one in the future.)
+
+## The implementation and optimization of `.use`
+
+An expression `x.use`, where `x` has type `T`, requires that `T: Use`. However,
+`x.use` does not always invoke `Use::do_use(x)`; in some cases the compiler can
+optimize away a use.
+
+If `x` is statically known to be dead, the compiler will move `x` rather than
+using it. This allows functions to write `x.use` without concern for whether
+it's the last usage of `x` (e.g. when passing `x.use` to a series of
+functions). Much like a trailing comma, this allows every usage to be
+symmetric, making it easy to compare uses or add more uses afterwards. (This
+optimization also means we should generally not lint on a final use of `x.use`,
+such as we currently do with the clippy lint `redundant_clone`.)
+
+If `x` is not statically known to be dead, but *all* of the following
+conditions are met, the compiler *may* elide an `x.use` and use `&x` instead:
+- The compiler can statically see that `x` outlives the result of `x.use`,
+- The compiler can statically see that the result of `x.use` is only accessed
+  via shared reference (including methods with `&self`, dereferences via
+  `Deref`, other invocations of `.use`, `use ||` closures, or `async use`
+  blocks). Effectively, these conditions mean that the user could theoretically
+  have refactored the code to use `&x` rather than `x.use`.
+
+An example of these elisions:
+
+```rust
+fn f(obj: Arc<Object>) {
+    g(obj.use); // `g` takes `Arc<Object>`
+    h(use || obj.method()); // `Object::method` here takes `&self`
+}
+
+fn main() {
+    let obj = Arc::new(Object::new());
+    f(obj.use);
+    f(obj.use);
+    f(obj.use);
+    g(obj.use);
+}
+```
+
+If every invocation of `.use` or `use ||` here resulted in a call to
+`.do_use()`, this program would call `.do_use()` 10 times (and have 11 `Arc`s
+to drop); however, the compiler can elide all the uses in `main` and have `f`
+work with the original Arc, resulting in only 6 calls to `.do_use()` (and only
+7 `Arc`s to drop). When an object is used repeatedly, such as in a loop, this
+can result in many elisions over the course of the program, and lower
+contention for atomics.
+
+If a user has an unusual `Use` type for which they wish to avoid these
+potential elisions, they can call `.do_use()` directly. (In a future edition of
+Rust, `Use` will be in the prelude, allowing calls to `.do_use()` without
+importing anything.)
```

Comment:
If a user wants to prevent elision of `do_use()`, and `do_use()` is defined to always do the same thing as `Clone::clone`, what situations would make the user want to call `.do_use()` over `.clone()`? Why does `do_use` need to be a separate stable[^1] method instead of just using `T: Use` bounds and calling `t.clone()`?

I suppose you could have a case where you have `bxarc: Box<Arc<T>>` or similar, where `bxarc.use` would give `Arc<T>`, so to replace `bxarc.do_use()` you would have to do something like `Arc::clone(&bxarc)` instead of `bxarc.clone()`, but this is already an example that is not necessarily super clear to begin with, so adding the explicitness might be better anyway, especially if the user wants to explicitly disable `.use` elision.

Perhaps "not have `do_use` be a separate method" could be added as an alternative/unresolved question?

[^1]: Considering only public usage, not potential compiler-internal usage of `do_use` in the desugaring of `.use` syntax that may benefit from it being separate from `Clone::clone`.

Reply:
@zachs18 That's a very good point. I've captured this as an alternative for now, but I agree that there's very little value in being able to call `.do_use()`, so I'm seriously considering just switching the RFC over to calling `Clone::clone` directly.

Reply:
@zachs18 I've updated the RFC to eliminate `Use::do_use`. Thank you!